My Site Dashboard: Refactor next post cells and remove unneeded code

hassaanelgarem opened the pull request:
Part of #18308

## Description
- This PR concludes the refactor 🎉 
- Refactored `BlogDashboardNextPostView` into a collection view cell `DashboardEmptyPostsCardCell`
- Added two cells inheriting from `DashboardEmptyPostsCardCell`
- Removed `DashboardPostCardCell` and `PostsCardViewController` completely
- Removed unnecessary code from multiple places
- Handles cases where cards needs to be reloaded to correctly show the current state

_PS: I know the lines changed are a LOT. However, ~500 of those are deleted files 😄_  

## Testing Instructions

Since the refactor is now done, please test that the dashboard behaves as expected.
- Sanity check on the dashboard in general:
    - Quick actions
    - Stats
    - Quick start
    - Loading state (Ghost)
    - Failure card
- A more detailed sanity check on posts cards. Some scenarios to consider:
    - Test drafts, scheduled, create next, create first cards are shown when expected
    - Test draft and scheduled lists are updated normally (Adding or removing posts)
    - Test drafts and scheduled cards are removed when all posts are removed (either by trashing or publishing)
    - Test showing drafts card after the first draft is created
    - Test showing scheduled card after the first scheduled post is created
    - Test switching from create first card to create next card when the first post is published   

## Regression Notes
1. Potential unintended areas of impact
Posts cards

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hassaanelgarem committed:
Refactor: Change `NextPostView` to be a cell and use it for relevant cards

hassaanelgarem committed:
Merge branch 'feature/dashboard-arch-refactor/posts-card-refactor' into feature/dashboard-arch-refactor/posts-card-refactor-2

hassaanelgarem committed:
Refactor: handle removing all posts from posts list card

hassaanelgarem committed:
Merge branch 'feature/dashboard-arch-refactor/posts-card-refactor' into feature/dashboard-arch-refactor/posts-card-refactor-2

hassaanelgarem committed:
Refactor: Add a cell class for each empty posts cell

hassaanelgarem committed:
Refactor: handle showing drafts/scheduled cards when new post is created

hassaanelgarem committed:
Refactor: reload cards if needed when a post is published

hassaanelgarem committed:
Remove: `DashboardPostsCardCell`

hassaanelgarem committed:
Remove: `PostsCardViewController`

hassaanelgarem committed:
Remove: unneeded code from `PostsCardViewModel`

hassaanelgarem committed:
Remove: unneeded code from `BlogDashboardViewController`

hassaanelgarem committed:
Fix: avoid showing empty posts list for a second before removing the cell

hassaanelgarem committed:
Add: documentation for `BlogDashboardViewModel` usage

momo-ozawa commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift:
```diff
@@ -100,6 +107,12 @@ class BlogDashboardViewModel {
 private extension BlogDashboardViewModel {
 
+    func registerNotifications() {
+        NotificationCenter.default.addObserver(self, selector: #selector(self.showDraftsCardIfNeeded), name: .newPostCreated, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(self.showScheduledCardIfNeeded), name: .newPostScheduled, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(self.showNextPostCardIfNeeded), name: .newPostPublished, object: nil)
```
[nit]
```suggestion
        NotificationCenter.default.addObserver(self, selector: #selector(showDraftsCardIfNeeded), name: .newPostCreated, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(showScheduledCardIfNeeded), name: .newPostScheduled, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(showNextPostCardIfNeeded), name: .newPostPublished, object: nil)
```

hassaanelgarem replied:
Fixed! 13459481be39ad98ac1b1800c6215e01ebebdafa

momo-ozawa reviewed (approved):
I did the general smoke test as well as the more detailed smoke tests and it works as expected!

Thanks for the refactor!

leandroalonso commented:
I made a bunch of tests and I found a few issues. I would split them into:

1) Important
2) Couldn't reproduce consistently
3) Minor issues

## Important

#### Jetpack site (role: collaborator)

I can consistently reproduce an issue in a Jetpack-connected site that I'm collaborator:

1. Write the first post
2. Save as draft
3. Delete it
4. Go back
5. The dashboard will show a blank "drafts card"

https://user-images.githubusercontent.com/7040243/163033614-4b6f4c51-1274-4235-89e3-2a07fe2f167b.mp4

At one moment, the next prompt didn't appear at all for me, even after updating it:

https://user-images.githubusercontent.com/7040243/163033676-994da29d-f3c7-46dc-8fde-ce92862847cf.mp4

#### Offline

1. Open a site with no drafts
2. Save your draft
3. The next prompt doesn't update

We cannot rely solely on the API to update the state of the cards because the user might create something offline. In theory, the code logic is correct, so I'm wondering what might be happening.

I also had this issue where it was showing me the drafts card and not the next post prompt, but I couldn't reproduce it consistently:

https://user-images.githubusercontent.com/7040243/163034942-50a8f002-66bd-4b2b-aa2b-5aa83cfd6fc5.mp4

#### Showing a deleted post while in a slow connection

I did a test with a .com site:

1. On a slow connection: created a draft for a site with no drafts
2. Deleted it
3. Returned to the dashboard
4. I was able to open the recently deleted post and edit it

This is one we should def address it. If a post is deleted it should disappear from the list right away, we can run into some nasty Core Data crashes by keeping a deleted object alive in memory.

For this one in specific, I think we should still keep the logic to show the next post prompt if the table view has zero results. I think it kinda clashes with the idea of having one cell per card but I don't have any other ideas. Unless maybe we can reload the cards from cache and we would know there aren't posts to display?

https://user-images.githubusercontent.com/7040243/163034990-3d973412-4b80-47f0-8ef1-dda5e8ac5720.mp4

## Couldn't reproduce consistently

When switching from the site above to a .com I was able to see the drafts card with the wrong height:

<img src="https://user-images.githubusercontent.com/7040243/163033860-baeb1d8c-ada8-4181-bfec-dc0d9e08542e.png" width="300">

I wonder if the cell was reused and kept the previous size or something like that.

## Minor issues

Most of those are regressions:

### Flickering

I consider this one minor but it bothers me a lot (I don't think most users will be switching sites frequently though). Comparing the old version and this version, the dashboard flickers most of the time when loading the cells (while before it was more instantly). The video below shows the current version vs the one in this branch (real device):

https://user-images.githubusercontent.com/7040243/163035444-bc1abb39-bf1b-4a56-a943-d3119e1837e7.MP4

I wonder if that it's because the view model is always being instantiated, which can be more time costly than the dequeuing strategy from before (in which changing sites pretty much just changed the fetched request and reconfigured the cells. But I'm not sure, thoughts?

### Loading message doesn't appear

I think this one you already created an issue for it.

### Syncing issues

This is an issue when creating a draft or scheduled from the API.

1. Open the app
2. Go to wordpress.com
3. Create a draft for your site
4. Go back to Menu and then back to Home
5. Sometimes it takes a while until the new post appear

You can delete and restore the post to check that the reflections sometimes doesn't appear.

(In the current `trunk` version this update occurs quite fast).

--

I still have on my list to test on iPhone 13 — given we found out a few issues last time with it.

leandroalonso commented:
@hassaanelgarem another thing: you might want to change the `RELEASE-NOTES` and add an internal note about the posts card with instructions to be tested during the call for testing.

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardEmptyPostsCardCell.swift:
```diff
@@ -83,19 +108,61 @@ class BlogDashboardNextPostView: UIView {
+
+private extension DashboardEmptyPostsCardCell {
+    func presentEditor() {
+        guard let blog = blog, let viewController = viewController else {
```
This code will reintroduce the editor delay issue, I recommend updating to the one here: https://github.com/wordpress-mobile/WordPress-iOS/pull/18332

hassaanelgarem replied:
Fixed! 0744783eb0c284a77c35fd24ad0250684309f291

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift:
```diff
@@ -30,8 +28,6 @@ class PostsCardViewModel: NSObject {
-
     private var syncing: (NSNumber?, BasePost.Status)?
 
     private weak var viewController: PostsCardView?
```
What do you think about renaming that to simple `view`?

hassaanelgarem replied:
Agree! Changed in 9db63f09c797f454eae1df486928e21fb3fb273a

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift:
```diff
@@ -291,6 +251,11 @@ extension PostsCardViewModel: NSFetchedResultsControllerDelegate {
             return
         }
+
+        guard updatePostsInfoIfNeeded() == false else {
```
I think this part here might be causing the deleted post to still show after being deleted.

hassaanelgarem replied:
This part basically detects when the fetchController has zero posts. When this is detected, we halt reloading the posts list table view and trigger the dashboard collection to reload all cards.

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift:
```diff
@@ -65,19 +65,19 @@ enum DashboardCard: String, CaseIterable {
 
-    private func shouldShowRemoteCard(apiResponse: BlogDashboardRemoteEntity?) -> Bool {
-        guard let apiResponse = apiResponse else {
+    private func shouldShowRemoteCard(postsInfo: BlogDashboardPostsInfo?) -> Bool {
```
I think this is too specific to posts, what do you think about returning the `apiResponse` here and having the `BlogDashboardPostsInfo` inside it?

The reason is because I can already see a case in which this will need to be changed:

1. Take into account the currently in-development Blogging Prompts card
2. We don't want to show this card and the "nextPost" at the same time
3. Here we would check (from the API) if the blogging prompts card should appear
4. However... all we have is the posts info.

Wdyt?

hassaanelgarem replied:
I completely agree with you!
The posts info approach was actually problematic for other reasons as well. I completely removed it while fixing an issue here 141b43042e3425b9eabecb292f764b39aabe76e8

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift:
```diff
@@ -130,6 +143,42 @@ private extension BlogDashboardViewModel {
+
+    // In case a post is published and create_first card
+    // is showing, we replace it with the create_next card.
+    @objc func showNextPostCardIfNeeded() {
```
Maybe we should add methods for when posts are deleted?

hassaanelgarem replied:
That's a good idea, but I don't think it will be as simple as that unfortunately. 
As mentioned in https://github.com/wordpress-mobile/WordPress-iOS/pull/18351#issuecomment-1098921097, I will tackle this in a different PR.

leandroalonso replied:
Hmmm, I was relying on the theory that maybe we could call `loadCardsFromCache` but I believe that in this case, the cached API response is that there are posts but they're not local, so it would try to load them, right?

It's indeed not that simple.

hassaanelgarem replied:
The posts aren't instantly removed from the card because `PostService` marks it as "trashed" only after the API call finishes. This is why it stays a while displayed in the card, because in the eyes of the fetchController it is still in CoreData.

The way `PostListViewController` handles this more gracefully is by keeping in memory a list of all "recently trashed posts". Recently trashed posts are posts that have been deleted but the delete API hasn't finished yet. For these, the ViewController displays the restore post cell.

Triggering a call to `loadCardsFromCache` will not fix this, because the post will still be in CoreData and returned by the fetch request.

What I'm planning to explore is to adjust the fetch request in a way to not return "recently trashed" posts. This should fix our problem. And if there's only one post and it gets deleted, this should automatically handle displaying the next post card.

leandroalonso reviewed (commented):
@hassaanelgarem thank you for tackling this! I think we're heading to a much better path than before. I left a few suggestions and I did a long test (you can see my comment above) in which I found a few regressions.

One additional thing, it looks like we're firing duplicated events for the posts:

```
022-04-12 16:37:13:684 WordPress[17142:331117] 🔵 Tracked: my_site_dashboard_card_shown <default_tab_experiment: site_menu, sub_type: draft, type: post>
2022-04-12 16:37:13:723 WordPress[17142:331117] 🔵 Tracked: my_site_dashboard_card_shown <default_tab_experiment: site_menu, sub_type: draft, type: post>
```

I guess it's because this doesn't contain the recently added entity to prevent that, so probably will be tackled in an upcoming PR. :)

hassaanelgarem committed:
Refactor: use cells to display loading and error views in posts card

momo-ozawa commented:
Adding a few  things I saw while testing today.

### Draft post doesn't load as a contributor on a Jetpack site

1. Switch to a site where you're a contributor for a Jetpack site / you have drafts
2. Tap on the posts quick action button and go to the drafts tab
3. Verify that you have drafts
4. 👀 Notice that the drafts card shows up but shows an "Unable to load posts right now" error
5. Tap on the draft post to retry loading
6. 👀 Notice that the draft card shows drafts for a second before displaying the "Unable to load posts right now" error


https://user-images.githubusercontent.com/6711616/163196923-24082641-586f-4b08-995a-36afcc222170.mp4


### Cards not loading from cache when offline

This is probably related to the offline issue mentioned above, but noting it here for completeness.

#### Drafts -> Drafts
1. Switch to a site with drafts (let's call this site A)
2. Switch to another site with drafts
3. Turn on the 100% loss profile on the network link conditioner
4. Switch back to site A
5. 👀 Observe that the draft card shows the ghostable tableview loading state, when it should display the cached drafts immediately

https://user-images.githubusercontent.com/6711616/163201279-3b9ab051-f011-45fa-847b-656fa235a7cf.mov


#### Drafts -> No posts (first post or next post)
1. Switch to a site with drafts (let's call this site A)
2. Switch to another site with no posts (either first post or next post)
3. Turn on the 100% loss profile on the network link conditioner
4. Switch back to site A
5. 👀 Observe that the draft card isn't shown and that the "Create your first post" or the "Create your next post" card is displayed instead

https://user-images.githubusercontent.com/6711616/163201943-14a46ad3-d66e-4d1b-8554-2a80ed2afa96.mov

hassaanelgarem committed:
Fix: enhance offline support

hassaanelgarem committed:
Fix: hide loading if local data exists

hassaanelgarem committed:
Fix: avoid reloading cells unnecessarily

hassaanelgarem committed:
Fix: tests build errors

hassaanelgarem committed:
Fix: modify error cell color in dark mode

hassaanelgarem committed:
Refactor: enhance loading display logic

hassaanelgarem committed:
Change: remove unnecessary self usage

hassaanelgarem committed:
Refactor: rename PostsCardViewModel.viewController to view

hassaanelgarem committed:
Refactor: rename updatePostsInfoIfNeeded to removeViewIfNeeded

hassaanelgarem committed:
Merge branch 'feature/dashboard-arch-refactor/base' into feature/dashboard-arch-refactor/posts-card-refactor-2

# Conflicts:
#	WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewController.swift

hassaanelgarem commented:
@leandroalonso
Thank you for the detailed test coverage!
Here are the updates I have on the issues/comments you've mentioned: 
- **Jetpack site (role: collaborator)**
    - I couldn't reproduce this using the same website. Not sure if I fixed it accidentally or if it was an API issue. Can you check this again? 🙏 
- **Offline support**
    - Fixed in 141b43042e3425b9eabecb292f764b39aabe76e8
- **Showing a deleted post while in a slow connection**
    - This one I was able to reproduce on `trunk` as well 😬 
    - Since it's unrelated to this PR and given the complexity of fixing this, I will tackle this in another PR instead
-  **Drafts card with the wrong height**
    - Fixed in 228975b25d6461d4e6ee812774d034464b5c2718
- **Flickering**
    - Fixed (accidentely 😄) in 141b43042e3425b9eabecb292f764b39aabe76e8
- **Loading message doesn't appear**
    - Fixed in 228975b25d6461d4e6ee812774d034464b5c2718
- **Syncing issues**
    - I was able to reproduce this. Given the size of the PR already, I will tackle this in another PR.
- **Firing duplicated events for the posts**
    - Fixed in 0744783eb0c284a77c35fd24ad0250684309f291
- **Internal release note**
    - Yes, great point! I'll add a note in the final PR that will merge the feature branch into `trunk`

hassaanelgarem commented:
@momo-ozawa 
Thank you for surfacing these issues!
Here are the updates I have on the issues you've mentioned: 
- **Draft post doesn't load as a contributor on a Jetpack site**
    - I couldn't reproduce this using the same website. Not sure if I fixed it accidentally or if it was an API issue. Can you check this again? 🙏 
- **Cards not loading from cache when offline**
    - **Drafts -> Drafts**: fixed in 67508285837ef76a5d9149ce040728854cf6c54f
    - **Drafts -> No Posts**: fixed in 141b43042e3425b9eabecb292f764b39aabe76e8

momo-ozawa commented:
@hassaanelgarem 👋 

> Draft post doesn't load as a contributor on a Jetpack site 

I can still consistently reproduce this. 

On the Jetpack Complete site, I make sure to have at least one draft. Then I delete the app and do a fresh install for the latest commit, and the "Unable to load posts right now" error is always displayed. Maybe the fresh install is the key to triggering this state?

If I delete all drafts, the "First post" card is displayed as expected. But if I try making a draft post, the "Unable to load posts right now" error is displayed again.

> Cards not loading from cache when offline 

I can confirm that the Drafts -> Drafts and Drafts -> No Posts cases are fixed. Thank you!

hassaanelgarem commented:
@momo-ozawa This is very weird! I followed the exact same steps and couldn't reproduce this. Tried this 3 times now.

https://user-images.githubusercontent.com/25306722/163379605-0b910296-fcf4-4d89-8f92-8e9ed04eb258.mov

I tried as well opening the Jetpack Complete site as the initial site after login, but same thing.

momo-ozawa commented:
@hassaanelgarem Super strange. 

I tested on a real device using the same steps and wasn't able to reproduce this. 

Meanwhile on a simulator I'm consistently able to reproduce this issue -- I deleted derived data, reset my simulator, did a fresh install, tried on a different simulator. For reference I'm using an iOS 15.2 simulator. 

Perhaps this is a simulator issue?

leandroalonso commented:
@momo-ozawa can you put breakpoints on `syncAuthors` and `showLoadingFailureError` to see from where the error is coming from?

If that is displayed to you it means that 1) sync authors are failing (but it shouldn't be triggered in the first place) and 2) that requesting the posts is failing.

Only difference I can think of between simulator and device is the proxy.

leandroalonso commented:
@hassaanelgarem thanks for tackling almost all of that. 😅

**Jetpack site (role: collaborator)**

I can't reproduce that either anymore!

**Offline support**

Can attest it's fixed!

**Showing a deleted post while in a slow connection**

Ok! Curiously, I tested on a slow network, and... it worked as expected this time. 🤔

**Drafts card with the wrong height**

Wasn't able to reproduce it again!

**Flickering**

🎊

**Loading message doesn't appear**

👍

**Firing duplicated events for the posts**

🚨As far as I can see the event for drafts and scheduled cards are not being fired.

leandroalonso commented:
@hassaanelgarem regarding the issue Momo mentioned:

> Draft post doesn't load as a contributor on a Jetpack site

I think there's a piece of logic which I'm not sure is taken into account with the refactor: if there are local posts (like in the video she shared) we should not display the error message. The error message should be displayed only if it failed to load and there are no posts to display.

Perhaps the `createSnapshot` which checks the `currentState` is forcing the error view.

leandroalonso commented on WordPress/Classes/Services/PostCoordinator.swift:
```diff
@@ -243,6 +243,9 @@ class PostCoordinator: NSObject {
             if uploadedPost.isScheduled() {
                 self?.notifyNewPostScheduled()
             }
+            else if uploadedPost.isPublished() {
```
[nit]

```suggestion
            } else if uploadedPost.isPublished() {
```

hassaanelgarem replied:
Fixed in bfe27af8d36df3282b0247f5bb65d916edae9f7f

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsListCardCell.swift:
```diff
@@ -73,43 +74,8 @@ class DashboardPostsListCardCell: UICollectionViewCell, Reusable {
-
     func trackPostsDisplayed() {
-        WPAnalytics.track(.dashboardCardShown, properties: ["type": "post", "sub_type": status.rawValue])
+        BlogDashboardAnalytics.shared.track(.dashboardCardShown, properties: ["type": "post", "sub_type": status.rawValue])
```
This isn't being called anywhere I think.

hassaanelgarem replied:
Fixed in eff29bb13d78bbfc7a83a2be4b8d5dce2b9716f3

leandroalonso reviewed (commented):
@hassaanelgarem almost there! 🎊

To summarize my many comments, I think those are the issues we need to fix before merging:

* Drafts and scheduled not being tracked
* Error message shouldn't appear if there's posts to display

momo-ozawa commented:
@leandroalonso @hassaanelgarem 

> Only difference I can think of between simulator and device is the proxy.

I think this is the reason. The "unable to load post right now" error only happens for me when I turn on AutoProxxy. If I disable it, everything works as expected.

> can you put breakpoints on syncAuthors and showLoadingFailureError to see from where the error is coming from?
> 
> If that is displayed to you it means that 1) sync authors are failing (but it shouldn't be triggered in the first place) and 2) that requesting the posts is failing.

I'm seeing the error because `syncAuthors` is failing.  

The proxy seems to affect whether `blog.isAdmin` evaluates to true or false. When AutoProxxy is enabled,`syncAuthors` is triggered because `blog.isAdmin` evaluates to true.

momo-ozawa commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardPostsParser.swift:
```diff
@@ -19,19 +19,25 @@ class BlogDashboardPostsParser {
+                if !posts.hasDrafts, localDraftsCount > 0 {
+                    posts["draft"] = Array(repeatElement([:], count: localDraftsCount))
+                }
             }
```
What do you think of getting rid of the else block and combining into one block?

```suggestion
            // If drafts are synced OR
            // If drafts are not synced and the cards API returns zero posts
            // depend on local data
            if blog.dashboardState.draftsSynced || (!posts.hasDrafts && localDraftsCount > 0) { 
                posts["draft"] = Array(repeatElement([:], count: localDraftsCount))
            }
```

hassaanelgarem replied:
I initially split them to enhance readability. But I think that comment now clears everything up 👍 
Fixed in 652675f769f219409308953b9b46b03eec564d25

momo-ozawa commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardPostsParser.swift:
```diff
@@ -19,19 +19,25 @@ class BlogDashboardPostsParser {
+                if !posts.hasScheduled, localScheduledCount > 0 {
+                    posts["scheduled"] = Array(repeatElement([:], count: localScheduledCount))
+                }
             }
```
Similar to the comment above

```suggestion
            // If scheduled posts are synced OR
            // If scheduled posts are not synced and the cards API returns zero posts
            // depend on local data
            if blog.dashboardState.scheduledSynced || (!posts.hasScheduled && localScheduledCount > 0) {
                posts["scheduled"] = Array(repeatElement([:], count: localScheduledCount))
            }
```

hassaanelgarem replied:
Fixed in 652675f769f219409308953b9b46b03eec564d25

hassaanelgarem committed:
Fix: track posts card when local data is available

hassaanelgarem committed:
Fix: only show error state if no posts are available locally

hassaanelgarem committed:
Fix: coding style of else if

hassaanelgarem committed:
Refactor: combine if statement in one block

hassaanelgarem commented:
@momo-ozawa @leandroalonso Thank you for getting to the bottom of the Jetpack issue! 🙏 

@leandroalonso Re your summary of final issues:
- Drafts and scheduled not being tracked
    - Fixed in eff29bb13d78bbfc7a83a2be4b8d5dce2b9716f3
- Error message shouldn't appear if there's posts to display
    - Fixed in f1b7646e31c24edb67270d708b638a846910d1bd

leandroalonso reviewed (approved):
Tested the two remaining issues, all good! Thank you so much for working on that, Hassaan!

hassaanelgarem committed:
Fix: reorder tests execution
